packages/jsii/lib/transforms/deprecation-warnings-transformer.ts

```diff
@@ -0,0 +1,618 @@
+import {
+  Assembly,
+  ClassType,
+  InterfaceType,
+  isClassOrInterfaceType,
+  Stability,
+} from '@jsii/spec';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+
+import * as bindings from '../node-bindings';
+import {
+  fullyQualifiedName,
+  isDeprecated,
+  isInternal,
+  isPublic,
+} from './utils';
+
+const WARNING_FUNCTION_NAME = 'printJsiiDeprecationWarnings';
+const FILE_NAME = '.warnings.jsii.js';
+
+type MethodLikeDeclaration =
+  | ts.MethodDeclaration
+  | ts.ConstructorDeclaration
+  | ts.GetAccessorDeclaration
+  | ts.SetAccessorDeclaration;
+
+interface Warning {
+  readonly elementName: string;
+  readonly message: string;
+  readonly path?: string;
+}
+
+/*
+
+General idea:
+
+When the JSII assembly is created, DeprecatedWarningInjector.process should be
+invoked so that it can compute and store general information about the module,
+such as deprecated types and class and interface declarations. It also creates a
+new Javascript file with a function that prints warnings.
+
+At emit time, DeprecatedWarningsTransformer.transform is invoked for each source
+file and then recursively for each node in the AST. We are interested in two
+types of nodes: class declarations and method-like declarations (proper methods,
+constructors and accessors).
+
+For classes, we navigate up in the type hierarchy looking for deprecated types.
+If one is found, a warning for that will added to each method of the class. For
```

Comment:
And I'm also not sure we need to do that. We would already get the deprecation warning when the class is constructed; what would additional deprecation warnings add every time we then call a member method on the class?

On the other hand, for STATIC methods we probably DO need to do this.

Reply:
I'm somewhat half-convinced the behavior of the assembler should already be to push deprecation markers down to subtypes, if applicable (so that if a method is deprecated in a superclass, it also counts as deprecated in a subclass). 

Not actually sure if that is true or not... but especially for consistency with the `deprecated-remover`, I would hope we don't do complex type analysis here and trust the analysis and markings done by the `assembler` already. We should centralize those kinds of decisions if we can.